### PR TITLE
C1: DeviceActivity interval validation

### PR DESCRIPTION
**Pull request by mnbf9rca**
## Summary
- Fixes #212 by clamping 24h timer intervals at the shared DeviceActivity chokepoint, removing the 1440 slider snap point, and rejecting 1440-minute Shortcuts starts with corrected copy.
- Fixes #228 by rejecting sub-15-minute combined schedule windows at validation time and reshaping stop-only schedule anchors before 00:15.
- Implements epic #263 bundle C1 only; no sub-15-minute enforcement/fallback mechanism was added.

## Maintainer Decisions
- #228A: reject at save via TriggerConfigurationModel.validate().
- #228B: reshape the internal stop-only interval anchor.
- Break-timer sub-15 behavior deferred to #214.

## Test Plan
- [x] xcodebuild test -project FamilyFoqos.xcodeproj -scheme FamilyFoqos -destination 'platform=iOS Simulator,id=B9E4A679-BDF3-4541-A59F-DA4BE21F80ED' | xcpretty
- [x] swift-format lint --recursive .

Fixes #212
Fixes #228
Refs #263

<h3>Greptile Summary</h3>

This PR implements bundle C1 of the DeviceActivity interval-validation epic: it fixes a 24-hour timer collapse bug (#212) by clamping `getTimeIntervalStartAndEnd` to 1439 minutes, fixes sub-15-minute stop-only schedule windows (#228) by reshaping the anchor and adding save-time validation, and removes the 1440 snap point from the duration slider.

- **`DeviceActivityLimits`** centralises the two framework-imposed constants (`minimumIntervalMinutes = 15`, `maximumTimerMinutes = 1439`) so all enforcement points share a single source of truth.
- **`stopScheduleInterval`** detects stop times before 00:15 and shifts the internal anchor one minute past the stop time, creating a ~1439-minute cross-midnight window that DeviceActivity will honour.
- **`TriggerConfigurationModel.validate()`** adds a save-time gate for schedule windows < 15 minutes using the new `TriggerValidator.scheduleWindowMinutes` modular-arithmetic helper, with a correct `window > 0` guard to avoid duplicate errors alongside the existing same-time check.

<h3>Confidence Score: 4/5</h3>

Safe to merge; the core interval-clamping and window-validation logic is correct and well-tested across all meaningful edge cases.

The three hardcoded duration strings in IntentError, StartProfileIntent, and the test assertions are the only rough edges — they can silently drift if DeviceActivityLimits constants change. All runtime logic (clamp, modular window math, stop-only anchor reshaping) is correct, and the test suite is thorough.

Foqos/Intents/IntentError.swift and Foqos/Intents/StartProfileIntent.swift carry hardcoded limit strings; FoqosTests/ScheduleWindowValidationTests.swift uses a hardcoded substring in its assertions.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| Foqos/Utils/DeviceActivityLimits.swift | New file — single source of truth for DeviceActivity interval limits (min 15 min, max 1439 min). Clean and well-documented. |
| Foqos/Utils/DeviceActivityCenterUtil.swift | Adds upper-bound clamp in getTimeIntervalStartAndEnd (fixes #212) and new stopScheduleInterval function for stop-only schedule anchoring (fixes #228). Logic and edge cases are correct. |
| Foqos/Models/TriggerConfigurationModel.swift | Adds sub-15-minute schedule window validation using scheduleWindowMinutes; correctly guards window > 0 to avoid duplicating the existing same-time error. |
| Foqos/Models/TriggerValidator.swift | Adds scheduleWindowMinutes static method using modular arithmetic; correctly handles both same-day and cross-midnight windows. |
| Foqos/Intents/IntentError.swift | Updated durationOutOfRange copy to '23 hours 59 minutes' but the string is hardcoded rather than derived from DeviceActivityLimits constants, creating a drift risk. |
| Foqos/Intents/StartProfileIntent.swift | Updated intent description copy to '23 hours 59 minutes' but similarly hardcoded rather than derived from DeviceActivityLimits. |
| FoqosTests/ScheduleWindowValidationTests.swift | Good coverage of scheduleWindowMinutes math and TriggerConfigurationModel integration, but assertions use hardcoded 'at least 15 minutes' strings rather than the DeviceActivityLimits constant. |
| FoqosTests/StopScheduleIntervalTests.swift | Covers stopScheduleInterval edge cases (midnight, 00:10, 00:15, 09:00) thoroughly; all assertions match the expected wrapping behavior. |
| FoqosTests/TimerIntervalTests.swift | Adds tests for 1440m clamp and multiples-of-day clamp; correctly verifies the non-zero-length invariant post-clamp. |
| FoqosTests/TimerDurationSnapTests.swift | New unit tests for snappedDuration static function, including a defense-in-depth case with a bad snap point above max. |
| FoqosTests/StrategyManagerBackgroundTests.swift | Adds test verifying that exactly 1440 minutes is now rejected with durationOutOfRange from startSessionFromBackground. |
| Foqos/Components/Strategy/TimerDurationView.swift | Removes 1440 snap point, extracts snappedDuration as a testable static method with max clamping. Logic is correct and now unit-tested. |
| Foqos/Utils/StrategyManager.swift | Updates duration bounds check to use DeviceActivityLimits constants; now correctly rejects exactly 1440 minutes. Error message matches the updated copy in IntentError. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[User / Shortcut sets duration] --> B{duration < 15 or > 1439?}
    B -- yes --> C[Throw durationOutOfRange\nError shown to user]
    B -- no --> D[Save profile]

    D --> E[scheduleTimerActivity]
    E --> F{hasV2StartSchedule?}
    F -- yes + stop schedule --> G[intervalStart = startTime\nintervalEnd = stopTime]
    F -- yes, no stop --> H[intervalEnd = startTime - 1 min\n~1439 min window]
    F -- no --> I[Legacy path]

    D --> J[scheduleStopActivity]
    J --> K{hasScheduledStop and NOT hasScheduledStart?}
    K -- no --> L[Remove stop-only activity]
    K -- yes --> M[stopScheduleInterval\nstopHour:stopMinute]
    M --> N{stopMinuteOfDay < 15?}
    N -- yes --> O[anchor = stopMinute + 1\n~1439 min cross-midnight window]
    N -- no --> P[anchor = 00:00\nwindow = stopMinuteOfDay minutes]
    O --> Q[startMonitoring]
    P --> Q

    D --> R[TriggerConfigurationModel.validate]
    R --> S[scheduleWindowMinutes\nmodular arithmetic]
    S --> T{window > 0 and window < 15?}
    T -- yes --> U[Error: window must be >= 15 min\nBlocks save]
    T -- no --> V[Valid config]

    style C fill:#f99,stroke:#c00
    style U fill:#f99,stroke:#c00
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[User / Shortcut sets duration] --> B{duration < 15 or > 1439?}
    B -- yes --> C[Throw durationOutOfRange\nError shown to user]
    B -- no --> D[Save profile]

    D --> E[scheduleTimerActivity]
    E --> F{hasV2StartSchedule?}
    F -- yes + stop schedule --> G[intervalStart = startTime\nintervalEnd = stopTime]
    F -- yes, no stop --> H[intervalEnd = startTime - 1 min\n~1439 min window]
    F -- no --> I[Legacy path]

    D --> J[scheduleStopActivity]
    J --> K{hasScheduledStop and NOT hasScheduledStart?}
    K -- no --> L[Remove stop-only activity]
    K -- yes --> M[stopScheduleInterval\nstopHour:stopMinute]
    M --> N{stopMinuteOfDay < 15?}
    N -- yes --> O[anchor = stopMinute + 1\n~1439 min cross-midnight window]
    N -- no --> P[anchor = 00:00\nwindow = stopMinuteOfDay minutes]
    O --> Q[startMonitoring]
    P --> Q

    D --> R[TriggerConfigurationModel.validate]
    R --> S[scheduleWindowMinutes\nmodular arithmetic]
    S --> T{window > 0 and window < 15?}
    T -- yes --> U[Error: window must be >= 15 min\nBlocks save]
    T -- no --> V[Valid config]

    style C fill:#f99,stroke:#c00
    style U fill:#f99,stroke:#c00
```

</a>

<sub>Reviews (1): Last reviewed commit: ["fix(#228): reshape stop-only interval an..."](https://github.com/mnbf9rca/family-foqos/commit/19f38d464e1e3d24b3cf6d6d777338abea081fb5) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41954886)</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

**Comment by mnbf9rca** [comment]
Code review requested before merge per AGENTS.md. Please review C1 implementation against docs/superpowers/plans/2026-07-05-c1-deviceactivity-interval-validation.md, especially scope guard: validation/surfacing only and no sub-15-minute enforcement fallback.

File: Foqos/Components/Strategy/TimerDurationView.swift [hunk 1]
```diff
@@ -18,7 +18,7 @@ struct TimerDurationView: View {
   private let largeIncrement: Double = 15
 
   // Common snap points (in minutes)
-  private let snapPoints: [Double] = [15, 30, 45, 60, 90, 120, 180, 240, 360, 480, 720, 1440]
+  private let snapPoints: [Double] = [15, 30, 45, 60, 90, 120, 180, 240, 360, 480, 720]
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/Components/Strategy/TimerDurationView.swift:21]: Fixed in e13ebd3: the slider min/max now derive from DeviceActivityLimits.

File: Foqos/Models/TriggerConfigurationModel.swift [hunk 2]
```diff
@@ -63,6 +63,22 @@ final class TriggerConfigurationModel: ObservableObject {
     {
       errors.append("Start and stop times can't be the same")
     }
+    if startTriggers.schedule && stopConditions.schedule,
+      let start = startSchedule, let stop = stopSchedule,
+      start.isActive, stop.isActive
+    {
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/Models/TriggerConfigurationModel.swift:70]: Fixed in e13ebd3: the same-time validation now requires both schedules to be active, with a regression test for inactive matching schedules.

File: FoqosTests/ScheduleWindowValidationTests.swift [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    let model = makeModel(startHour: 23, startMinute: 55, stopHour: 0, stopMinute: 5)
+    model.validate()
+    XCTAssertTrue(model.validationErrors.contains { $0.contains("at least 15 minutes") })
+  }
+
+  func testGivenFifteenMinuteWindow_WhenValidating_ThenNoWindowError() {
+    let model = makeModel(startHour: 9, startMinute: 0, stopHour: 9, stopMinute: 15)
+    model.validate()
+    XCTAssertFalse(model.validationErrors.contains { $0.contains("at least 15 minutes") })
+  }
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply FoqosTests/ScheduleWindowValidationTests.swift:79]: Fixed in 95c1026: the schedule-window tests now build the expected error fragment from DeviceActivityLimits.minimumIntervalMinutes.

File: Foqos/Intents/StartProfileIntent.swift [hunk 4]
```diff
@@ -17,7 +17,7 @@ struct StartProfileIntent: AppIntent {
   nonisolated(unsafe) static var title: LocalizedStringResource = "Start Family Foqos Profile"  // SAFETY: AppIntents requires static var; immutable after init
 
   nonisolated(unsafe) static var description = IntentDescription(  // SAFETY: AppIntents requires static var; immutable after init
-    "Start a Family Foqos blocking profile. Optionally specify a timer duration in minutes (15-1440)."
+    "Start a Family Foqos blocking profile. Optionally specify a timer duration between 15 minutes and 23 hours 59 minutes."
   )
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/Intents/StartProfileIntent.swift:21]: Leaving this literal intentionally. The approved C1 plan explicitly calls out StartProfileIntent as a LocalizedStringResource-backed string that must stay a string literal for string-catalog extraction; the values are framework-fixed and manually synchronized here.

File: Foqos/Intents/IntentError.swift [hunk 5]
```diff
@@ -16,7 +16,7 @@ enum IntentError: Error, CustomLocalizedStringResourceConvertible {
     case .sessionAlreadyActive:
       "A session is already active."
     case .durationOutOfRange:
-      "Duration must be between 15 and 1440 minutes."
+      "Duration must be between 15 minutes and 23 hours 59 minutes."
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/Intents/IntentError.swift:19]: Leaving this literal intentionally. The approved C1 plan explicitly calls out IntentError as a LocalizedStringResource-backed string that must stay a string literal for string-catalog extraction; the values are framework-fixed and manually synchronized here.

[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/Intents/IntentError.swift:19]: Leaving this literal intentionally. The approved C1 plan explicitly requires the IntentError LocalizedStringResource text to remain a string literal for string-catalog extraction; the values are framework-fixed and manually synchronized.

File: Foqos/Intents/StartProfileIntent.swift [hunk 6]
```diff
@@ -17,7 +17,7 @@ struct StartProfileIntent: AppIntent {
   nonisolated(unsafe) static var title: LocalizedStringResource = "Start Family Foqos Profile"  // SAFETY: AppIntents requires static var; immutable after init
 
   nonisolated(unsafe) static var description = IntentDescription(  // SAFETY: AppIntents requires static var; immutable after init
-    "Start a Family Foqos blocking profile. Optionally specify a timer duration in minutes (15-1440)."
+    "Start a Family Foqos blocking profile. Optionally specify a timer duration between 15 minutes and 23 hours 59 minutes."
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply Foqos/Intents/StartProfileIntent.swift:20]: Leaving this literal intentionally. The approved C1 plan explicitly requires the StartProfileIntent LocalizedStringResource text to remain a string literal for string-catalog extraction; the values are framework-fixed and manually synchronized.
